src/core/parser.ml

```diff
@@ -320,409 +23,120 @@ let add_prefix : string -> string -> string = fun p s ->
-  | _refl_                      -> Pos.in_pos _loc P_tac_refl
-  | _sym_                       -> Pos.in_pos _loc P_tac_sym
-  | i:{_:_focus_ nat_lit}       -> Pos.in_pos _loc (P_tac_focus(i))
-  | _why3_ s:string_lit?        -> Pos.in_pos _loc (P_tac_why3(s))
-  | q:query                     -> Pos.in_pos _loc (P_tac_query(q))
-  | _fail_                      -> Pos.in_pos _loc P_tac_fail
-  | _solve_                     -> Pos.in_pos _loc P_tac_solve
-
-(** [proof_end] is a parser for a proof terminator. *)
-let parser proof_end =
-  | _end_   -> P_proof_end
-  | _admit_ -> P_proof_admit
-  | _abort_ -> P_proof_abort
-
-let parser assoc =
-  | EMPTY   -> Assoc_none
-  | "left"  -> Assoc_left
-  | "right" -> Assoc_right
-
-(** [config] parses a single configuration option. *)
-let parser config =
-  | "builtin" s:string_lit "≔" qid:qident ->
-      P_config_builtin(s,qid)
-  | "prefix" p:float_lit s:string_lit "≔" qid:qident ->
-      let unop = (s, p, qid) in
-      sanity_check _loc_s s;
-      Prefix.add unops s unop;
-      P_config_unop(unop)
-  | "infix" a:assoc p:float_lit s:string_lit "≔" qid:qident ->
-      let binop = (s, a, p, qid) in
-      sanity_check _loc_s s;
-      Prefix.add binops s binop;
-      P_config_binop(binop)
-  | "declared" id:string_lit ->
-      sanity_check _loc_id id;
-      Prefix.add declared_ids id id;
-      P_config_ident(id)
-  | "unif_rule" r:unif_rule ->
-      P_config_unif_rule(r)
-  | "quantifier" qid:qident ->
-      P_config_quant(qid)
-
-let parser proof =
-  _begin_ ts:tactic* e:proof_end -> (ts, Pos.in_pos _loc_e e)
-
-(** [!require mp] can be used to require the compilation of a module [mp] when
-    it is required as a dependency. This has the effect of importing notations
-    exported by that module. The value of [require] is set in [Compile], and a
-    reference is used to avoid to avoid cyclic dependencies. *)
-let require : (Path.t -> unit) Stdlib.ref = ref (fun _ -> ())
-
-(** [do_require pos path] is a wrapper for [!require path], that takes care of
-    possible exceptions. Errors are reported at given position [pos],  keeping
-    as much information as possible in the error message. *)
-let do_require : Pos.pos -> p_module_path -> unit = fun loc path ->
-  let path = List.map fst path in
-  let local_fatal fmt =
-    let fmt = "Error when loading module [%a].\n" ^^ fmt in
-    parser_fatal loc fmt Path.pp path
+module Lp : PARSER = struct
+  let parse_lexbuf : string -> Sedlexing.lexbuf -> Syntax.ast =
+    fun fname lexbuf ->
+    Sedlexing.set_filename lexbuf fname;
+    let parse =
+      MenhirLib.Convert.Simplified.traditional2revised LpParser.command
+    in
+    let lexer = LpLexer.lexer lexbuf in
+    let cmds = ref [] in
+    try
+      while true do
+        let cmd = parse lexer in
+        cmds := cmd :: !cmds;
+      done;
+      assert false (* Should raise end of file before *)
+    with
+    | End_of_file -> List.rev !cmds
```

Comment:
Why do you parse command by command and then have to reverse? Why not defining an entry in LpParser for handling a list of commands?
Idem for Dk below.
Remark: later we could improve all this by interleaving parsing and handling: indeed we could handle each command once it is parser; in this case, there is no need for building a list of commands.

Reply:
From memory, parsing a list of commands raised an error in the parser generation (something like end of stream conflict). I'll check that again.

Reply:
Parsing and handling are interleaved in 88e97d5 using a stream rather than a list. Does this suit you?